### PR TITLE
Update build

### PR DESCRIPTION
**Pull request by tomdee**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,30 +1,77 @@
...
+
+$(NODE_CONTAINER_BIN_DIR)/%: calico_node/%.py
+	-docker run -v `pwd`:/code --rm \
+	 calico/build:latest \
+	 sh -c 'pyinstaller -ayF --distpath $(@D) $< && chown $(shell id -u):$(shell id -g) -R $(@D)'
+
+$(NODE_CONTAINER_BIN_DIR)/calico-felix:
+	-docker rm -f calico-felix
+	# TODO - update the felix image when https://github.com/projectcalico/calico/pull/1111 is merged
+	docker create --name calico-felix calico/felix:pyinstaller
```
Comment (djosborne) [inline Makefile]:
```
# Latest felix binaries are stored in automated builds of calico/felix.
# To get them, we pull that image, then copy the binaries out to our host
```

File: Makefile [hunk 2]
```diff
@@ -1,30 +1,77 @@
...
+	docker cp calico-felix:/code/. $(@D)
+
+$(NODE_CONTAINER_BIN_DIR)/confd:
```
Comment (djosborne) [inline Makefile]: `# Get the confd binary`

File: Makefile [hunk 3]
```diff
@@ -1,30 +1,77 @@
...
+	curl -L https://github.com/projectcalico/confd/releases/download/v0.10.0-scale/confd.static -o $@
+	chmod +x $@
+
+$(NODE_CONTAINER_BIN_DIR)/bird:
```
Comment (djosborne) [inline Makefile]: `# Get bird binaries`

File: Makefile [hunk 4]
```diff
@@ -38,22 +85,41 @@ dist/calicoctl: $(CALICOCTL_FILE) birdcl gobgp
...
+
+gobgp:
+	docker pull osrg/gobgp
+	docker run \
+	-v `pwd`:/code \
+	--entrypoint=sh \
+	osrg/gobgp \
+	-c 'cp /go/bin/gobgp /code'
+
+simple-binary: $(CALICOCTL_FILE) birdcl gobgp
```
Comment (djosborne) [inline Makefile]: I thought `birdcl` and `gobgp` were `calico/node` deps, not `calicoctl` deps

File: calico_node/Dockerfile [hunk 5]
```diff
@@ -14,29 +14,23 @@
...
-  rm -f glibc.apk glibc-bin.apk && \
-  rm -rf /var/cache/apk/*
+RUN apk --no-cache add wget ca-certificates libgcc && \
+    wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://raw.githubusercontent.com/sgerrand/alpine-pkg-glibc/master/sgerrand.rsa.pub && \
+    wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.23-r3/glibc-2.23-r3.apk && \
+    wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.23-r3/glibc-bin-2.23-r3.apk && \
+    apk add glibc-2.23-r3.apk glibc-bin-2.23-r3.apk && \
+    /usr/glibc-compat/sbin/ldconfig /lib /usr/glibc/usr/lib && \
+    apk del wget ca-certificates && \
+    rm -f glibc-2.23-r3.apk glibc-bin-2.23-r3.apk
```
Comment (tomdee) [inline calico_node/Dockerfile]: An apk is like a deb. We download it, then install it. There's no need to leave the file lying around wasting space.

File: calico_node/build.sh [hunk 6]
```diff
@@ -1,46 +0,0 @@
...
-curl -L https://github.com/projectcalico/calico-bird/releases/download/v0.1.0/bird6 -o /sbin/bird6
-curl -L https://github.com/projectcalico/calico-bird/releases/download/v0.1.0/birdcl -o /sbin/birdcl
-chmod +x /sbin/*
-
-# Install Felix and libcalico
-pip install git+https://github.com/projectcalico/calico.git
-pip install git+https://github.com/projectcalico/libcalico.git
-
-# Output the python library list
-pip list > libraries.txt
```
Comment (tomdee) [inline calico_node/build.sh]: I has no meaning now. There is no pip and no pythong

Comment (tomdee) [inline calico_node/build.sh]: *It

File: Makefile [hunk 7]
```diff
@@ -1,30 +1,77 @@
...
+NODE_CONTAINER_BINARIES=startup allocate-ipip-addr calico-felix bird calico-bgp-daemon confd
+
+calico-node.tar: $(NODE_CONTAINER_CREATED)
+	docker save --output $@ calico/node:latest
+
+# Get docker2aci from https://github.com/appc/docker2aci
+calico-node-latest.aci: calico-node.tar
+	docker2aci $<
+
+$(NODE_CONTAINER_CREATED): $(NODE_CONTAINER_DIR)/Dockerfile  $(addprefix $(NODE_CONTAINER_BIN_DIR)/,$(NODE_CONTAINER_BINARIES))
```
Comment (tomdee) [inline Makefile]: Yes, that's the whole point of a Makefile. If I do two builds in a row the second is instant. If I change a file, then make knows what to rebuild

File: Makefile (hunk 4 above)
Comment (tomdee) [inline Makefile]: You are incorrect, let's discuss offline.

File: Makefile [hunk 8]
```diff
@@ -38,22 +93,41 @@ dist/calicoctl: $(CALICOCTL_FILE) birdcl gobgp
 	 calico/build:latest \
 	 pyinstaller calicoctl.spec -ayF
 
-simple-binary:
+birdcl:
```
Comment (djosborne) [inline Makefile]: `gobgp` and `birdcl` are dropped into project root on build. Can we volume mount pwd/dist instead?

**Comment by djosborne** [comment]
I've got two untracked files after building that should get added to the gitignore
```
        allocate-ipip-addr.spec
        calico_node/filesystem/bin/
```

File: Makefile [hunk 9]
```diff
@@ -38,22 +93,41 @@ dist/calicoctl: $(CALICOCTL_FILE) birdcl gobgp
...
+	wget -N https://github.com/projectcalico/calico-bird/releases/download/v0.1.0/birdcl
+	chmod +x birdcl
+
+gobgp:
```
Comment (djosborne) [inline Makefile]: Might as well add the last few generic comments:
```
# Get gobgp binary
```

File: Makefile (hunk 8 above)
Comment (djosborne) [inline Makefile]:
```
# Get birdcl binary
```

**Comment by djosborne** [comment]
Since some of these steps don't have a comment the line above, `make help` isn't very useful anymore. I could consider it out of scope, but maybe raise an issue to ensure all relevant build steps appear in  `make help` output?

File: Makefile [hunk 10]
```diff
@@ -1,30 +1,85 @@
...
+# Build binary from python files, e.g. startup.py or allocate-ipip-addr.py
+$(NODE_CONTAINER_BIN_DIR)/%: calico_node/%.py
+	-docker run -v `pwd`:/code --rm \
+	 calico/build:latest \
+	 sh -c 'pyinstaller -ayF --distpath $(@D) $< && chown $(shell id -u):$(shell id -g) -R $(@D)'
+
+# Get felix binaries
+$(NODE_CONTAINER_BIN_DIR)/calico-felix:
+	-docker rm -f calico-felix
+	# TODO - update the felix image when https://github.com/projectcalico/calico/pull/1111 is merged
```
Comment (djosborne) [inline Makefile]: 1111 was merged

**Comment by djosborne** [comment]
Couple remaining things. Then probably LGTM but I'd like to take one final look just to see how they were resolved

**Comment by tomdee** [comment]
I've added the gitignore lines and removed the TODO now that PR 1111 is merged.

I'm not going to fix up all the help commands or add more comments - I think this should be merged as is.

**Pull request merged**